x-pack/plugins/observability_solution/apm_data_access/server/services/get_host_names/index.ts

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { estypes } from '@elastic/elasticsearch';
+import { rangeQuery } from '@kbn/observability-plugin/server';
+import { RegisterParams } from '../register_services';
+
+export interface ServicesHostNamesParams {
+  query: estypes.QueryDslQueryContainer;
+  from: number;
+  to: number;
+  limit: number;
+}
+
+export function createGetHostNames(params: RegisterParams) {
+  return async ({ from: timeFrom, to: timeTo, query, limit }: ServicesHostNamesParams) => {
+    const { apmIndices, esClient } = await params.getResourcesForServices();
+
+    const esResponse = await esClient.search({
+      index: [apmIndices.metric, apmIndices.transaction, apmIndices.span, apmIndices.error],
```

Comment:
This will be slow, because you're aggregating over all data. Even if it's just a single service, you'll run into performance issues. What would be more efficient here is if you figure out what data source to query first (tx metrics or raw events, 1m vs 10m and 60m), and then execute it.

Reply:
that would be the same thing that the [`get_document_sources` does](https://github.com/elastic/kibana/blob/main/x-pack/plugins/observability_solution/apm/server/lib/helpers/get_document_sources.ts). We'd have to move that to `apm_data_access`. Is that your idea?

Reply:
yeah, how involved would that be?

Reply:
it's a relatively big change. A few things would have to be moved (could be more):

[`document_type`](https://github.com/elastic/kibana/blob/main/x-pack/plugins/observability_solution/apm/server/lib/helpers/create_es_client/document_type.ts), [`get_document_sources`](https://github.com/elastic/kibana/blob/main/x-pack/plugins/observability_solution/apm/server/lib/helpers/get_document_sources.ts) and the types they depend on.

Reply:
ok, sounds fine to do it in a follow-up but I'd like to make it happen, to set a good first example

Reply:
yeah, makes sense. I'm going to get a sense of how much effort it would be. it might as well be doable in this PR.